### PR TITLE
Adjusting `ServiceDiscovererEvent` contract

### PR DESCRIPTION
**Pull request by chemicL**
Motivation:

Current API of `ServiceDiscovererEvent` is capable of representing the
boolean status of the host being either available or not available. As
seen in the implementation of `RoundRobinLoadBalancer`, the resulting
states from the Service Discovery system can have more meanings than a
simple boolean. To prepare for enhanecements, the contract has been
adjusted to allow more states and make it possible for further additions
if those are needed by particular implementations.

Modifications:

- Deprecated `ServiceDiscovererEvent#isAvailable` method and added a
  default implementation to allow removal in child classes,
- Added `ServiceDiscoveryStatus` class,
- Added `ServiceDiscovererEvent#status` method with default
  implementation to prevent breaking the existing implementations,
- Replaced existing uses of `isAvailable` with comparisons of `status`
  against the constants in `ServiceDiscoveryStatus`,
- Added TODOs in order to reflect the need to adapt the codebase to
  support non-binary statuses.

Result:

`ServiceDiscovererEvent`s are prepared for more advanced, non-binary
statuses. An example is the now introduced `EXPIRED` event, which can have a
different meaning than `UNAVAILABLE`. If the `ServiceDiscoverer` decides
to emit such a status, the subscribers can provide additional
functionality in that case.

### CODE REVIEW

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java
Comment (idelpivnitskiy) [reply servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java:222]: I see it as a risk if RRLB does not follow new SD contract. Releasing new API will motivate users to migrate. If they use all 3 events, they won't see desired behavior.

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java
Comment (chemicL) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java]: Actually, implemented as part of this PR, PTAL.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java
Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java:167]: In the meantime, the SD retry strategy has been deprecated.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java
Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java:273]: All occurrences now use `status.equals`.

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java
Comment (chemicL) [reply servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java]: As discussed, I implemented handling of EXPIRED status here and also in the Dns layer.

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoveryStatus.java
Comment (chemicL) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscoveryStatus.java]: I had some concerns but in the end decided to go with your suggestion. PTAL.

File: servicetalk-client-api-internal/src/main/java/io/servicetalk/client/api/internal/ServiceDiscovererUtils.java
Comment (chemicL) [reply servicetalk-client-api-internal/src/main/java/io/servicetalk/client/api/internal/ServiceDiscovererUtils.java]: I actually removed this method altogether and assume relying on `Status.equals(...)` method.

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java
Comment (chemicL) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java:756]: `Status.equals(...)` is used instead.

**Comment by chemicL** [comment]
Seems the build failure might be related to caching, as the error should not appear with the current state of the file mentioned.

```
/home/runner/work/servicetalk/servicetalk/servicetalk-client-api/src/main/java/io/servicetalk/client/api/DefaultServiceDiscovererEvent.java:37: warning - Tag @link: can't find DefaultServiceDiscovererEvent(Object, Status) in io.servicetalk.client.api.DefaultServiceDiscovererEvent

> Task :servicetalk-client-api:javadoc FAILED
```

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java
Comment (idelpivnitskiy) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java:756]: It's used in all other places except this one

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java
Comment (idelpivnitskiy) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java:273]: This one still uses `==`

File: servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DefaultDnsClientTest.java
Comment (idelpivnitskiy) [inline servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DefaultDnsClientTest.java:84]: Leftover?

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java
Comment (idelpivnitskiy) [inline servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java:137]: Thanks for describing in this comment how users should migrate. However, when we deprecate something, we should not kill functionality. With the new SD contract, we can still make this feature work as expected: apply a `map` operator for SD publisher and depending on `eagerConnectionShutdown` flag, flip `EXPIRED` -> `UNAVAILABLE` or `UNAVAILABLE` -> `EXPIRED`.

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java
Comment (idelpivnitskiy) [inline servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java:221]: Consider using static imports for all statuses

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererObserver.java
Comment (idelpivnitskiy) [inline servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DnsServiceDiscovererObserver.java:106]: This will actually break all uses of old API, because `DefaultResolutionResult` only implements `nMissing()`. We should swap `default` implementations in this interface: default `nUnavailable()` should delegate to `nMissing()`, while `nMissing()` should throw `UnsupportedOperationException`. It will be safe for our implementation, and will provide smooth migration if anyone else implements this interface.

**Review by idelpivnitskiy — COMMENTED** [review]
The PR is marged as `breaking-change`, but is expected to land in 0.41. I don't see any breaking changes after the last comments will be addressed. Consider running `scripts/japicmp.sh` to make sure and remove the lable. 

LGTM after comments addressed. Feel free to ping if you want me to double check before merging.

**Comment by chemicL** [comment]
> Seems the build failure might be related to caching, as the error should not appear with the current state of the file mentioned.
> 
> ```
> /home/runner/work/servicetalk/servicetalk/servicetalk-client-api/src/main/java/io/servicetalk/client/api/DefaultServiceDiscovererEvent.java:37: warning - Tag @link: can't find DefaultServiceDiscovererEvent(Object, Status) in io.servicetalk.client.api.DefaultServiceDiscovererEvent
> 
> > Task :servicetalk-client-api:javadoc FAILED
> ```

As a matter of fact, it seems to fail only in case of Java 8. Possibly caused by https://bugs.openjdk.java.net/browse/JDK-8031625. I submitted a follow up using the fully qualified name and that fixes the issue.

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java
Comment (chemicL) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java:756]: That is indeed ironic. Thanks for paying attention!

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java
Comment (chemicL) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java:132]: True, but when a reference to a status is held not in context of an event, losing that contextual wrapper could confuse users when debugging and investigating, I prefer to keep it.

Comment (chemicL) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java:132]: Instead of your suggestion, I will go with adding a `public String getName()` method to `Status`. I believe with that `ServiceDiscovererEvent` can just use `Status#getName` in `toString` while preserving the desired properties described above.

File: servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java
Comment (chemicL) [reply servicetalk-dns-discovery-netty/src/main/java/io/servicetalk/dns/discovery/netty/DefaultDnsClient.java:144]: Makes sense.

File: servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DefaultDnsClientTest.java
Comment (chemicL) [reply servicetalk-dns-discovery-netty/src/test/java/io/servicetalk/dns/discovery/netty/DefaultDnsClientTest.java:84]: Yep.

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java
Comment (chemicL) [reply servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java]: Agreed, killing the feature was a mistake. I brought it back. However, there's one drawback with it. The users won't be able to handle three states separately. If LB was able to distinguish expired from unavailable and provide unavailable events as well as expired, it will not be possible to turn off the mapping. Now users have to coordinate between the state the LB emits (in case of the built-in implementation it's DNS) and the flag here. It's impossible to swap the new default `EXPIRED` to `UNAVAILABLE` and expect everything to work if the previous setting of eagerConnectionShutdown was the default `false`.  The new state was just introduced, perhaps it's worth considering a non-binary setting here -> "If not set, accept the raw state and act upon it".

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java
Comment (idelpivnitskiy) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java:139]: There is a high change that a `Status` instance will be passed to a logger. It's known that status has no other fields except `name` => not necessary to add `name=` part. Name of the class is also less useful. Debugger always shows the type, if users log they usually know what they log and do it with some other context info, ST core does not print it without an event. 
For the constants like this one, all other final classes just print it's main meaning. Have a look at `HttpRequestMethod`, `HttpResponseStatus`, `HttpProtocolVersion`.

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java
Comment (idelpivnitskiy) [inline servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java:52]: Thanks for the updated javadoc here. Would be nice to provide a short summary what statuses are supported by this implementation, before you go into "The created instances have the following behaviour" details.

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java
Comment (idelpivnitskiy) [inline servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java:190]: Remove "TODO"

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java
Comment (idelpivnitskiy) [inline servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java:284]: `@Nonnull` is not necessary as it's added automatically by `@ElementsAreNonnullByDefault`

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java
Comment (idelpivnitskiy) [reply servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java]: Good point! 🎖️ 
I like suggestion of making an internal `eagerConnectionShutdown` as `Boolean` and modifying the status only if it was explicitly set by users.

**Review by idelpivnitskiy — APPROVED** [review]

File: servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java
Comment (chemicL) [reply servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java:284]: Autogenerated after extract method...

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java
Comment (chemicL) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/ServiceDiscovererEvent.java:139]: Ok, I am convinced with the "enum" part ;-)

**Comment by chemicL** [comment]
0.41: ad81b3fdb8114a63fe349a9e9c7b008c1bcf17aa

**Pull request merged**
